### PR TITLE
[AMDGPU] Assert if stack grows downwards.

### PR DESCRIPTION
**Pull request by easyonaadit**

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp [hunk 1]
```diff
@@ -1181,8 +1181,8 @@ bool AMDGPURegisterBankInfo::applyMappingDynStackAlloc(
 
   // Guard in case the stack growth direction ever changes with scratch
   // instructions.
-  if (TFI.getStackGrowthDirection() == TargetFrameLowering::StackGrowsDown)
-    return false;
+  assert(TFI.getStackGrowthDirection() == TargetFrameLowering::StackGrowsUp &&
+         "Stack grows upwards for AMDGPU\n");
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:1185]: No need for \n at the end.

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 2]
```diff
@@ -4041,17 +4041,15 @@ SDValue SITargetLowering::lowerDYNAMIC_STACKALLOCImpl(SDValue Op,
   Chain = SP.getValue(1);
   MaybeAlign Alignment = cast<ConstantSDNode>(Tmp3)->getMaybeAlignValue();
   const TargetFrameLowering *TFL = Subtarget->getFrameLowering();
-  unsigned Opc =
-      TFL->getStackGrowthDirection() == TargetFrameLowering::StackGrowsUp
-          ? ISD::ADD
-          : ISD::SUB;
+  assert(TFL->getStackGrowthDirection() == TargetFrameLowering::StackGrowsUp &&
+         "Stack grows upwards for AMDGPU\n");
```
Comment (cdevadas) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:4045]: Ditto.

**Review by arsenm — APPROVED** [review]

**Review by cdevadas — APPROVED** [review]

**Review by pravinjagtap — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 3]
```diff
@@ -4041,17 +4041,15 @@ SDValue SITargetLowering::lowerDYNAMIC_STACKALLOCImpl(SDValue Op,
...
+  assert(TFL->getStackGrowthDirection() == TargetFrameLowering::StackGrowsUp &&
+         "Stack grows upwards for AMDGPU");
 
   SDValue ScaledSize = DAG.getNode(
       ISD::SHL, dl, VT, Size,
       DAG.getConstant(Subtarget->getWavefrontSizeLog2(), dl, MVT::i32));
 
   Align StackAlign = TFL->getStackAlign();
-  Tmp1 = DAG.getNode(Opc, dl, VT, SP, ScaledSize); // Value
+  Tmp1 = DAG.getNode(ISD::ADD, dl, VT, SP, ScaledSize); // Value
```
Comment (s-barannikov) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:4052]: (not related to this PR, but I thought it was worth reporting)

If I'm reading correctly, the logic is broken for the case when the object alignment is greater than the stack alignment.
The `ISD::AND` below may result in an address that points below the current SP, possibly into a previously allocated object.

Consider: `SP = 40, StackAlign = 8, ScaledSize = 16, (Scaled)Alignment = 32`
```
Tmp1 = ADD 40, 16 = 56
Tmp1 = AND 56, -32 = 32
```

The new address (32) is less than the previous value of the stack pointer (40).

Comment (easyonaadit) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:4052]: Hi,
There are 2 things to consider here:
1. For AMDGPU, the Stack grows upwards (low addr to high addr)
2. The code here is based on the default implementation, which itself doesn't seem to adjust for a growing up stack.

This patch is actually just a part of fixing the logic flaw you pointed out. There is a fix in progress, you could follow it here: 
https://github.com/llvm/llvm-project/pull/119822

Comment (easyonaadit) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:4052]: Other than scaling for wave size, the code changes in https://github.com/llvm/llvm-project/pull/119822 serve as a generic implementation for a growing up stack. And there are test cases along with it as well. Could this be replicated?

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIISelLowering.cpp:4052]: The only other target that uses positive stack growth is NVPTX

**Pull request merged**